File: src/main/kotlin/com/kotlin/数据类型/类型转换例子.java
package com.kotlin.数据类型;

/**
 * Created by Lin Change on 2017-07-09.
 */
public class 类型转换例子 {
    public static void main(String[] args) {
        Parent parent = new Child();
        System.out.println(((Child)parent).getName());

        parent = new Parent();
        System.out.println(((Child)parent).getName()); //将抛出类型转换异常

        if (parent instanceof Child) {
            System.out.println(((Child) parent).getName());
        }
    }
}

File: src/main/kotlin/com/kotlin/数据类型/类型转换例子.kt
package com.kotlin.数据类型

/**
 * Created by Lin Change on 2017-07-09.
 */
fun main(args: Array<String>) {
//    val parent: Parent = Child()
    val parent: Parent = Parent()
//    val child: Child? = parent as Child //进行类型转换，如不成功则抛出异常
    val child: Child? = parent as? Child //进行类型转换，如不成功返回null
    println(child)

    if (parent is Child) {
        println(parent.name)
    }

    val string: String? = "Hello"
    if (string is String) { //智能转换1
        println(string.length)
    }
    if (string != null) { //智能转换2
        println(string.length)
    }
    println(string!!.length) //告诉编译器当前已确定字符串不为空
}
